### PR TITLE
Optional JSON-encoded query parameters

### PR DESCRIPTION
**Pull request by Gabrola**
Now the library can truly claim to offer "Fully typed RPC-like client"

I think the main point of discussion would be the opinionated approach of the selective encoding format of root-level strings.

The implementation, as is, will not JSON encode (i.e. surround with double quotes) a root-level string unless it clashes with a JSON reserved word or is a number.

```typescript
const query = { foo: 'foo', nested: { bar: 'bar' }  };
// ?foo=foo&nested={"bar":"bar"}

const query = { foo: 'true' };
// ?foo="true"

const query = { foo: true };
// ?foo=true

const query = { foo: '123' };
// ?foo="123"

const query = { foo: 123 };
// ?foo=123
```

I think it's nice because this leaves the URL looking like regular query strings in the majority of use cases. This is also how swagger encodes root-level strings even if we specify the MIME type as `application/json`

Screenshot 2022-12-29 062946

### CODE REVIEW

**Comment by arivera-xealth** [comment]
this is sick!

**Comment by oliverbutler** [comment]
I really like this approach! I'm going to review now, however I notice there is a merge conflict @Gabrola - If it's simple I'll try fix it now.

File: libs/ts-rest/core/src/lib/query.ts [hunk 1]
```diff
@@ -1,13 +1,43 @@
...
-export const convertQueryParamsToUrlString = (query: unknown) => {
-  const queryString = encodeQueryParams(query);
+export const convertQueryParamsToUrlString = (query: unknown, json = false) => {
+  const queryString = json
+    ? encodeQueryParamsJson(query)
+    : encodeQueryParams(query);
   return queryString?.length > 0 ? '?' + queryString : '';
 };
 
+export const encodeQueryParamsJson = (query: unknown) => {
```
Comment (oliverbutler) [inline libs/ts-rest/core/src/lib/query.ts:14]: Wonderfully simple implementation, I love it ❤️

**Review by oliverbutler — APPROVED** [review]

**Comment by Gabrola** [comment]
Cheers mate! ♥

**Pull request merged**